### PR TITLE
Make compare_sourcelists produce a single combined pdf with all the plots and statistical summaries

### PR DESCRIPTION
**Pull request by dulude**
1) drizzlepac/devutils/comparison_tools/compare_sourcelists.py:
- Plotting updates:
     - Instead of generating many individual difference plot .pdf files, compare_sourcelists.py now combines all the individual difference plots into a single multi-page .pdf file. 
     - Each difference plot is accompanied by a second page containing a statistical summary. 
     - The ordering of the plots/summaries in the combined pdf file is the exact same as the ordering when plots are displayed to the screen.
- Changed test pass/fail criteria:
     - All non-flag value comparisons: To pass, the sigma-clipped mean percent difference must be less then or equal to 5% 
     - Flag values: To pass, no more than 5% of of all matched sources can have flagging differences
- Spruced up primary and subroutine docstrings
- Other assorted cosmetic tweaks

2) drizzlepac/hapsequencer.py:
- All comparision_plots.pdf multipage plot pdf files generated for given total product object are now further combined into one single pdf file.
- compare_sourcelists() is now run computing percent differences (as in differences as a percentage of the sigma-clipped mean HLA sourcelist value) instead of absolute differences

3) setup.py
- added PyPDF2

**NOTE**
This update requires the installation of the "[PyPDF2](https://pypi.org/project/PyPDF2/)" python library to function.

### CODE REVIEW

**Comment by stsci-hack** [comment]
@dulude :  There are a few instances where variables are defined and not used, but otherwise, the following reported Flake8 issues identify ways in which this code needs to meet the standards for code style.  These need to be corrected prior to merging.  In the meantime, I am in the process of reviewing the logic and functional aspects of this PR. 

**Flake8 errors for `hapsequencer.py`:**
Error	Flake8	E265 — block comment should start with '# '	554:5
Error	Flake8	E231 — missing whitespace after ','	560:68
Error	Flake8	E231 — missing whitespace after ','	560:87
Error	Flake8	E231 — missing whitespace after ','	561:64
Error	Flake8	E231 — missing whitespace after ','	561:80
Error	Flake8	E226 — missing whitespace around arithmetic operator	561:96
Error	Flake8	E226 — missing whitespace around arithmetic operator	561:100
Error	Flake8	E261 — at least two spaces before inline comment	561:118
Error	Flake8	E231 — missing whitespace after ','	562:53
Error	Flake8	E261 — at least two spaces before inline comment	562:70
Error	Flake8	E261 — at least two spaces before inline comment	566:19
Error	Flake8	E231 — missing whitespace after ','	569:79
Error	Flake8	E261 — at least two spaces before inline comment	570:83
Error	Flake8	E225 — missing whitespace around operator	575:36
Error	Flake8	E226 — missing whitespace around arithmetic operator	575:62
Error	Flake8	E231 — missing whitespace after ','	580:92
Error	Flake8	E261 — at least two spaces before inline comment	584:103
Error	Flake8	E231 — missing whitespace after ','	595:100
Error	Flake8	E231 — missing whitespace after ','	595:149
Error	Flake8	E231 — missing whitespace after ','	595:164
Error	Flake8	E231 — missing whitespace after ','	595:181
Error	Flake8	E231 — missing whitespace after ','	595:227
Error	Flake8	E261 — at least two spaces before inline comment	599:41
Error	Flake8	E262 — inline comment should start with '# '	599:42
Error	Flake8	E231 — missing whitespace after ','	601:73
Warning	Flake8	F841 — local variable 'return_status' is assigned to but never used	595:17


**Flake8 errors for `compare_sourcelists.py`:**
Error	Flake8	E722 — do not use bare except'	677:5
Error	Flake8	E722 — do not use bare except'	1153:9
Error	Flake8	E722 — do not use bare except'	1160:9
Warning	Flake8	F841 — local variable 'p1' is assigned to but never used	314:9
Warning	Flake8	F841 — local variable 'p2' is assigned to but never used	315:9
Warning	Flake8	F841 — local variable 'file_handles' is assigned to but never used	1272:5
Warning	Flake8	F841 — local variable 'titleSwapDict_cooOld' is assigned to but never used	1190:5

**Comment by dulude** [comment]
Further development regarding the refactoring of compare_sourcelists.py detailed in issue #556.

File: drizzlepac/devutils/comparison_tools/compare_sourcelists.py [hunk 1]
```diff
@@ -300,135 +399,166 @@ def computeLinearStats(matchedRA,plotGen,diffMode,plot_title,plotfile_prefix,ver
...
+    intersVal = 3
 
+    diffRA = matchedRA[1, :] - matchedRA[0, :]  # simple difference
     clippedStats = sigma_clipped_stats(diffRA, sigma=sigVal, maxiters=intersVal)
     pct_1sig = (float(np.shape(np.where(abs(diffRA) <= clippedStats[2]))[1]) / float(np.shape(diffRA)[0])) * 100.0
+    sigma_percentages = []
+    for sig_val in [1.0, 2.0, 3.0]:
+        sigma_percentages.append((float(np.shape(np.where((diffRA >= (clippedStats[0] - sig_val * clippedStats[2])) & (diffRA <= (clippedStats[0] + sig_val * clippedStats[2]))))[1])/float(np.shape(diffRA)[0])) * 100.0)
+    #pct_3sig = ((float(np.shape(np.where(diffRA <= 3.0*clippedStats[2] + clippedStats[0]))[1]) + float(np.shape(np.where(diffRA >= clippedStats[0] - 3.0 * clippedStats[2]))[1]))/float(np.shape(diffRA)[0])) * 100.0
     pct_five = (float(np.shape(np.where(abs(diffRA) <= 5.0))[1]) / float(np.shape(diffRA)[0])) * 100.0
```
Comment (stsci-hack) [inline drizzlepac/devutils/comparison_tools/compare_sourcelists.py]: This computes the statistics on each set of values.  However, regardless of whether it is being provided RA in arcseconds or X in pixels or FLUX1 in counts, this line will always compare the difference to a constant value of 5.0.  I would suggest that this value be provided as an input to this routine to allow the calling routine to determine what limit is appropriate for the given values and their units.

**Comment by dulude** [comment]
These updates were incorporated into PR #571, making this PR obsolete.

File: drizzlepac/devutils/comparison_tools/compare_sourcelists.py [hunk 2]
```diff
@@ -300,135 +399,166 @@ def computeLinearStats(matchedRA,plotGen,diffMode,plot_title,plotfile_prefix,ver
...
+                if log_line != "\n":
+                    stat_text_blob += log_line + "\n"
+                else:
+                    stat_text_blob += "\n"
+            stat_text_blob += "\n" + timestamp + "\n"
+            stat_text_blob += plotfile_prefix
+            fig.text(0.5, 0.5, stat_text_blob, transform=fig.transFigure, size=10, ha="center", va="center",
+                     multialignment="left", family="monospace")
+            fig.savefig(plotFileName.replace(".pdf", "_stats.pdf"))
+            plt.close()
```
Comment (stsci-hack) [inline drizzlepac/devutils/comparison_tools/compare_sourcelists.py]: This section of code could be turned into a simple generic function with 'log_output_string_list' and 'plotFileName' as the inputs. The new function would not only be called here, but also from 'computeFlagStats' (line 339-354?) as well and limit the amount of duplicate code in this module.

**Pull request closed without merging**